compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs

```diff
@@ -728,25 +728,40 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
             _ => local_decl.source_info.span,
         };
 
+        // With ref-binding patterns, the mutability suggestion has to apply to
+        // the binding, not the reference (which would be a type error):
+        //
+        // `let &b = a;` -> `let &(mut b) = a;`
+        // or
+        // `fn foo(&x: &i32)` -> `fn foo(&(mut x): &i32)`
         let def_id = self.body.source.def_id();
-        let hir_id = if let Some(local_def_id) = def_id.as_local()
+        let is_ref_binding_patterns = if let Some(local_def_id) = def_id.as_local()
             && let Some(body_id) = self.infcx.tcx.hir().maybe_body_owned_by(local_def_id)
         {
             let body = self.infcx.tcx.hir().body(body_id);
-            BindingFinder { span: pat_span }.visit_body(body).break_value()
+
+            if let Some(hir_id) = (BindingFinder { span: pat_span }).visit_body(body).break_value()
+                && let hir::Node::Local(hir::Local {
+                    pat: hir::Pat { kind: hir::PatKind::Ref(_, _), .. },
+                    ..
+                }) = self.infcx.tcx.hir_node(hir_id)
+            {
+                true
+            } else {
+                body.params.iter().any(|param| {
+                    if let hir::Pat { kind: hir::PatKind::Ref(_, _), span, .. } = param.pat
+                        && *span == pat_span
+                    {
```

Comment:
Is it possible ot extend the `BindingFinder` to look into function parameters as well?

Reply:
> Is it possible ot extend the `BindingFinder` to look into function parameters as well instead of special-casing here?

It is indeed better to query through BindingFinder. 
Done, thank you